cmd/monitor/monitor.go

```diff
@@ -0,0 +1,179 @@
+	err := monitor.connectToValidators()
+	if err != nil {
+		log.Fatalf("Monitor exiting: couldn't connect to all validators: %s", err)
+	}
+
+	// make request for hvs to validators
+	monitor.requestHeightLogs()
+
+	// run accountability algorithm
+	output := monitor.runAccountabilityAlgorithm()
+	log.Println(output)
+}
+
+func (monitor *Monitor) runAccountabilityAlgorithm() string {
+	numValidators := len(monitor.Validators)
+
+	// lower bound on the number of faulty processes
+	minFaulty := (numValidators-1)/3 + 1 // f+1
+
+	// count the number of responses from different validators
+	responseCount := 0
+
+	// wait until the specified timer expires
+	timer := time.NewTicker(time.Duration(monitor.Timeout) * time.Second)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-timer.C:
+			// exit because timer expired
+			return timeoutStatus
+
+		case status := <-monitor.receiveChannel:
+
+			// check status received (new hvs or validator disconnected)
+			// if we have delivered at least f + 1 message logs, run the monitor algorithm
+			if status && monitor.accAlgorithm.HeightLogs.Length() >= minFaulty {
+
+				// run monitor and get faulty processes
+				monitor.accAlgorithm.IdentifyFaultyProcesses(uint64(numValidators), monitor.FirstDecisionRound, monitor.SecondDecisionRound)
+
+				// if we have at least f + 1 faulty processes, the algorithm completed
+				if monitor.accAlgorithm.FaultySet.Length() >= minFaulty {
+					return successfulStatus
+				}
+			}
+
+			// count the number of responses from validators
+			responseCount++
+
+			if responseCount == numValidators {
+				// exit because no new hvs will arrive and avoid waiting longer
+				return failStatus
+			}
+		}
+	}
+}
+
+// method to resolve processes addresses and store connection objects
+func (monitor *Monitor) connectToValidators() error {
+
+	if monitor.Validators == nil || len(monitor.Validators) == 0 {
+		return fmt.Errorf("error: no validators given")
+	}
+
+	// resolve validator addresses given and connect to them
+	for _, val := range monitor.Validators {
+		conn, err := connection.Connect(val)
+		if err != nil {
+			return fmt.Errorf("error while connecting to one of the validators given: %s", err)
+		}
+		monitor.connections = append(monitor.connections, conn)
+	}
+
+	return nil
+}
+
+// request async HeightVoteSets from validators
+func (monitor *Monitor) requestHeightLogs() {
+
+	// prepare packet to send
+	packet := &connection.Packet{Code: connection.HvsRequest, Height: monitor.Height}
+
+	// start goroutines to send message and wait for reply for each validator
+	for _, conn := range monitor.connections {
+
+		// channel used to close connection once hvs has been received and stop to send the request
+		validatorCloseChannel := make(chan bool)
+
+		// receive packets from validator
+		go monitor.receiveHvsFromValidator(conn, validatorCloseChannel)
+
+		// periodic send request to validator
+		go conn.PeriodicSend(packet, validatorCloseChannel, sendTimer)
```

Comment:
Monitor does not need to periodically request Height-Vote-Set, since we assume the reliable links.

Reply:
Yes I know that the model assumes perfect links between processes but real systems can have failures of any kind (process failures, network failures etc.) and the idea was to handle more case scenarios. Nevertheless, if the processes are correct and the network is reliable, the implementation behaves overall as the monitor algorithm specs describe; on the other hand, if there are some failures (some validator not responding back to the request, monitor not receiving some packets etc.) the monitor tries to handle the situation in the best way possible, always following the algorithm. We can better discuss this in the next meeting though and I can show you my reasoning.

Reply:
I agree with you that it makes more sense to use this logic in a real-systems scenario.